File: src/server/Adaptive.ReactiveTrader.Server.ReferenceDataWrite/Domain/CurrencyPair.cs
using Adaptive.ReactiveTrader.Contract.Events.Reference;
using Adaptive.ReactiveTrader.EventStore.Domain;

namespace Adaptive.ReactiveTrader.Server.ReferenceDataWrite.Domain
{
    public class CurrencyPair : AggregateBase
    {
        public override object Identifier => $"ccyPair-{Symbol}";
        private string Symbol { get; set; }
        private int PipsPosition { get; set; }
        private int RatePrecision { get; set; }
        private decimal SampleRate { get; set; }
        private string Comment { get; set; }
        public bool IsActive { get; private set; }
        private bool IsCreated { get; set; }

        public CurrencyPair()
        {
        }

        public CurrencyPair(string symbol, int pipsPosition, int ratePrecision, decimal sampleRate, string comment)
        {
            if (IsCreated)
            {
                return;
            }

            RaiseEvent(new CurrencyPairCreatedEvent(symbol, pipsPosition, ratePrecision, sampleRate, comment));
        }

        public void Activate()
        {
            if (IsActive || !IsCreated)
            {
                return;
            }

            RaiseEvent(new CurrencyPairActivatedEvent(Symbol));
        }

        public void Deactivate()
        {
            if (!IsActive || !IsCreated)
            {
                return;
            }

            IsActive = false;
            RaiseEvent(new CurrencyPairDeactivatedEvent(Symbol));
        }

        public void Apply(CurrencyPairCreatedEvent evt)
        {
            Symbol = evt.Symbol;
            PipsPosition = evt.PipsPosition;
            RatePrecision = evt.RatePrecision;
            SampleRate = evt.SampleRate;
            Comment = evt.Comment;
            IsCreated = true;
        }

        public void Apply(CurrencyPairActivatedEvent evt)
        {
            IsActive = true;
        }

        public void Apply(CurrencyPairDeactivatedEvent evt)
        {
            IsActive = false;
        }
    }
}

File: src/client/src/services/model/currencyPair.js
export default class CurrencyPair {
  _symbol:string;
  _base:string;
  _terms:string;
  _ratePrecision:number;
  _pipsPosition:number;
  _displayString:string;

  constructor(symbol:string, ratePrecision:number, pipsPosition:number) {
    this._symbol = symbol;
    this._ratePrecision = ratePrecision;
    this._pipsPosition = pipsPosition;

    this._base = symbol.substr(0 , 3);
    this._terms = symbol.substr(3 , 3);
    this._displayString = `${this._base}/${this._terms}`;
  }

  get symbol():string {
    return this._symbol;
  }

  get ratePrecision():number {
    return this._ratePrecision;
  }

  get pipsPosition():number {
    return this._pipsPosition;
  }

  get base():string {
    return this._base;
  }

  get terms():string {
    return this._terms;
  }

  get displayString() {
    return this._displayString;
  }
}
